### DISCUSSION TITLE
TypeRuntimeWiring vs CodeRegistry

### DISCUSSION THREAD

**Discussion by martinbonnin** [graphql-java/graphql-java#2777] [Q&A] [answered]
The [getting started](https://www.graphql-java.com/documentation/getting-started) page uses `TypeRuntimeWiring` to register data fetchers. I understand it can be used like this:

```kotlin
        RuntimeWiring.newRuntimeWiring()
            .type("Query") {
                it.dataFetcher("hello", { "from TypeRuntimeWiring" })
            }
            .build()
```

But other pages mention `CodeRegistry`:

```kotlin
        RuntimeWiring.newRuntimeWiring().codeRegistry(
            GraphQLCodeRegistry.newCodeRegistry()
                .dataFetcher(
                    FieldCoordinates.coordinates("Query", "hello"), DataFetcher {
                        "from CodeRegistry"
                    }
                )
                .build()
        ).build()
```

Are these the same two things? Is one preferred over the other?

**Comment by bbakerman** [comment, marked as answer]
They are related but different things and they pre-exist each other so there is cross over.

Once upon a time - each `GraphqlFieldDefinition` held its data fetcher.  `TypeRuntimeWiring` was used in schema generation to set a fields data fetcher and also remember you can programmatically create schema (outside `SchemeGenerator`) and that set them into the field direct.


This is bad for separating the runtime model of a schema from its code.  So `GraphQLCodeRegistry` was invented as a holder - again for SDL generated schema (from SDL text) and also for programmatically built ones.

So we end up with overlap here.  Its possible to do it direct with a code registry by name like you can in `TypeRuntimeWiring` as you point out.

However `TypeRuntimeWiring` does more - it wires in `EnumValuesProvider`, it also wires `TypeResolver` for interfaces and unions and a small thing is you can default the data fetcher used per type via `graphql.schema.idl.TypeRuntimeWiring.Builder#defaultDataFetcher`

All of this is run by `graphql.schema.idl.SchemaGenerator` and it combines it down to a single `GraphQLCodeRegistry` in the end.

So I would recommend using `RuntimeWiring` / `TypeRuntimeWiring` if you use SDL schema generation for consistency when you need to provide type resolvers and enum values etc...

**Reply by martinbonnin** [reply]
Thanks a lot for the detailed answer 🙏

**Reply by martinbonnin** [reply]
Follow up question: is there a specific reason there's no `RuntimeWiring.newRuntimeWiring(existingRuntimeWiring)` like there is `GraphQLCodeRegistry.newCodeRegistry(existingCodeRegistry)`?

**Reply by bbakerman** [reply]
No not great reason.   Just hasnt been done

**Reply by martinbonnin** [reply]
👍 mind if I give a shot at `RuntimeWiring.newRuntimeWiring(existingRuntimeWiring)` one of these days? We have a [workaround in our project](https://github.com/apollographql/federation-jvm/blob/master/graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/Federation.java#L164) but it'd feel better to have it as first party API

**Comment by bbakerman** [comment]
Sure... Go for it
